### ISSUE TITLE
After upgrading to .NET 8: Error in ProcessRequest: System.NotSupportedException: Serialization and deserialization of 'System.Reflection.MethodBase' instances is not supported.

### ISSUE THREAD

**Issue by sbuethe** [tylerjensen/ServiceWire#89]
When I upgrade my project from .NET 7 to .NET 8 an exception is thrown on the 32nd log entry always (after 31 successful runs; see log below). The client side throws a 'pipe is broken' exception.

When downgrading to .NET 7 everything works as expected again.

The app affected uses RPC over NamedPipes only and the method being invoked that fails has one byte[] parameter and void as return type: public void SetCapturedLastSlideBitmap(byte[] capturedBitmap)

The log of the default logger is as follows. Please note that entries 33 to 64 were logged after restarting the app.

2024-02-27T19:28:31.827	Debug	SyncInterface for ICongressUI in 6ms.
2024-02-27T19:28:47.300	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:29:02.599	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:29:18.848	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:29:35.084	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:30:26.240	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:30:46.370	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:31:02.368	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:31:18.454	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:31:41.985	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:31:57.500	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:32:17.117	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:32:32.681	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:32:48.099	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:33:03.818	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:33:23.479	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:33:39.262	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:33:54.732	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:34:10.216	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:34:25.750	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:34:41.375	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:35:01.010	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:35:16.861	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:35:32.164	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:35:48.346	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:36:04.312	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:36:20.259	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:36:40.151	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:36:55.577	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:37:11.427	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:37:34.874	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:37:41.979	Error	Error in ProcessRequest: System.NotSupportedException: Serialization and deserialization of 'System.Reflection.MethodBase' instances is not supported. Path: $.TargetSite.:: ---> System.NotSupportedException: Serialization and deserialization of 'System.Reflection.MethodBase' instances is not supported.::   at System.Text.Json.Serialization.Converters.UnsupportedTypeConverter`1.Write(Utf8JsonWriter writer, T value, JsonSerializerOptions options)::   at System.Text.Json.Serialization.JsonConverter`1.TryWrite(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)::   at System.Text.Json.Serialization.Metadata.JsonPropertyInfo`1.GetMemberAndWriteJson(Object obj, WriteStack& state, Utf8JsonWriter writer)::   at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1.OnTryWrite(Utf8JsonWriter writer, T value, JsonSerializerOptions options, WriteStack& state)::   at System.Text.Json.Serialization.JsonConverter`1.TryWrite(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)::   at System.Text.Json.Serialization.JsonConverter`1.WriteCore(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)::   --- End of inner exception stack trace ---::   at System.Text.Json.ThrowHelper.ThrowNotSupportedException(WriteStack& state, NotSupportedException ex)::   at System.Text.Json.Serialization.JsonConverter`1.WriteCore(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)::   at System.Text.Json.Serialization.Metadata.JsonTypeInfo`1.Serialize(Utf8JsonWriter writer, T& rootValue, Object rootValueBoxed)::   at System.Text.Json.Serialization.Metadata.JsonTypeInfo`1.SerializeAsObject(Utf8JsonWriter writer, Object rootValue)::   at System.Text.Json.JsonSerializer.WriteBytesAsObject(Object value, JsonTypeInfo jsonTypeInfo)::   at ServiceWire.DefaultSerializer.Serialize(Object obj, String typeConfigName)::   at ServiceWire.ParameterTransferHelper.SendParameters(Boolean useCompression, Int32 compressionThreshold, BinaryWriter writer, Object[] parameters)::   at ServiceWire.Host.ProcessInvocation(ZkSession session, BinaryReader binReader, BinaryWriter binWriter, Stopwatch sw)::   at ServiceWire.Host.ProcessRequest(Stream readStream, Stream writeStream)

2024-02-27T19:58:10.891	Debug	SyncInterface for ICongressUI in 5ms.
2024-02-27T19:58:26.605	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:58:41.970	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:58:57.622	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:59:14.034	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:59:29.519	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T19:59:48.684	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T20:00:03.687	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T20:00:19.285	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T20:00:43.199	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T20:00:58.316	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T20:01:18.324	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T20:01:33.473	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T20:01:48.967	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T20:02:03.983	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T20:02:23.031	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T20:02:38.515	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T20:02:53.931	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T20:03:09.813	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T20:03:24.814	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T20:03:39.992	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T20:03:59.215	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T20:04:14.877	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T20:04:30.296	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T20:04:45.997	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T20:05:02.429	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T20:05:17.977	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T20:05:37.681	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T20:05:53.563	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T20:06:09.145	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T20:06:33.060	Debug	SyncInterface for ICongressUI in 2ms.
2024-02-27T20:06:40.044	Error	Error in ProcessRequest: System.NotSupportedException: Serialization and deserialization of 'System.Reflection.MethodBase' instances is not supported. Path: $.TargetSite.:: ---> System.NotSupportedException: Serialization and deserialization of 'System.Reflection.MethodBase' instances is not supported.::   at System.Text.Json.Serialization.Converters.UnsupportedTypeConverter`1.Write(Utf8JsonWriter writer, T value, JsonSerializerOptions options)::   at System.Text.Json.Serialization.JsonConverter`1.TryWrite(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)::   at System.Text.Json.Serialization.Metadata.JsonPropertyInfo`1.GetMemberAndWriteJson(Object obj, WriteStack& state, Utf8JsonWriter writer)::   at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1.OnTryWrite(Utf8JsonWriter writer, T value, JsonSerializerOptions options, WriteStack& state)::   at System.Text.Json.Serialization.JsonConverter`1.TryWrite(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)::   at System.Text.Json.Serialization.JsonConverter`1.WriteCore(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)::   --- End of inner exception stack trace ---::   at System.Text.Json.ThrowHelper.ThrowNotSupportedException(WriteStack& state, NotSupportedException ex)::   at System.Text.Json.Serialization.JsonConverter`1.WriteCore(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)::   at System.Text.Json.Serialization.Metadata.JsonTypeInfo`1.Serialize(Utf8JsonWriter writer, T& rootValue, Object rootValueBoxed)::   at System.Text.Json.Serialization.Metadata.JsonTypeInfo`1.SerializeAsObject(Utf8JsonWriter writer, Object rootValue)::   at System.Text.Json.JsonSerializer.WriteBytesAsObject(Object value, JsonTypeInfo jsonTypeInfo)::   at ServiceWire.DefaultSerializer.Serialize(Object obj, String typeConfigName)::   at ServiceWire.ParameterTransferHelper.SendParameters(Boolean useCompression, Int32 compressionThreshold, BinaryWriter writer, Object[] parameters)::   at ServiceWire.Host.ProcessInvocation(ZkSession session, BinaryReader binReader, BinaryWriter binWriter, Stopwatch sw)::   at ServiceWire.Host.ProcessRequest(Stream readStream, Stream writeStream)

**Comment by tylerje** [comment]
.NET 8 client and host?

**Comment by tylerje** [comment]
Can you share the interface code for ICongressUI?

**Comment by tylerje** [comment]
I'd like to look at the interface code to try to replicate this. I have not experimented with .NET 8 yet. It's possible that the dynamic proxy is not correct in .NET 8.

**Comment by sbuethe** [comment]
Oh yes, client is on .NET 4.8.1

Full interface looks as follows:

    public interface ICongressUI
    {
        void SetCapturedFirstSlideBitmap(byte[] capturedBitmap, Int64 windowHandle);
        void SetCapturedLastSlideBitmap(byte[] capturedBitmap);
        TimerState GetTimerState();
        List<String> GetSpeakerList();
        void SetNextSpeaker();
        void SetPreviousSpeaker();
        void SetSpeakerIndex(int i);
    }
    
Client always alternately calls SetCapturedFirstSlideBitmap first (which always succeeds) and then SetCapturedLastSlideBitmap. I assume that only SetCapturedLastSlideBitmap fails, because it simply fails after a certain number of calls.

**Comment by tylerje** [comment]
Alternatively, try returning an int rather than void.

**Comment by sbuethe** [comment]
Unfortunately, returning int instead of void did not help. Same as before.

**Comment by tylerje** [comment]
Create a .NET 8 test client to see if the problem persists.

**Comment by sbuethe** [comment]
Just tested with a .NET 8 client to .NET 8 host and same behaviour.

**Comment by tylerje** [comment]
I will work on it but my availability is limited. If you find a fix, feel free to submit a PR.

**Comment by tylerje** [comment]
Are you closing the client and creating it with every call? Or do you keep the named pipe connection open, using only one client instance. I reread your original comment. I wonder if in .NET 8 there is a limit to the number of connections. That seems to be the most likely culprit. Can you change your code to use only one connection and reuse that connected client?

**Comment by sbuethe** [comment]
While the server remains open all the time, the client is actually closed after every other call - which unfortunately cannot be avoided since the apps containing the client code must be closed after the final bitmap is transmitted.

Indeed, a connection limit would actually explain why the requests always only fail after a fixed number of attempts.

Do you have any suggestion how I could best avoid exceeding the connection limit? I'm assuming that closing the pipe connection on the host resets the active number of connections, but what happens if a new client request comes in at just that time? I've implemented a retry logic in the meantime anyway, but it doesn't seem a very elegant solution to me.

And significantly increasing the number of allowed connections (if it's even possible) is probably not a good idea either.

**Comment by tylerje** [comment]
Try using TCP on local localhost? Named Pipes isn't designed for a relatively unlimited number of client connections, so I think you'll find TCP will work better for you.

**Comment by sbuethe** [comment]
Ok, I will modify it to use TCP and report back.

Many thanks for your support!

**Comment by sbuethe** [comment]
I modified my apps to use Tcp protocol instead of Named Pipes and unfortunately the problem persists.

The relevant part in the log now looks like this:

2024-03-01T20:07:36.565	Debug	ZkInitiation client username received: username
2024-03-01T20:07:36.565	Debug	ZkInitiation client Ephemeral received: +H/NU400d8Za0QyeQSPPCIGjwKR7Steb+m8oQ6k1iJE=
2024-03-01T20:07:36.565	Debug	ZkInitiation hash salt sent to client: kSw9ttPH/XEuxjKQgMaG5DMVWeswuHYfxoCea3++71U=
2024-03-01T20:07:36.565	Debug	ZkInitiation server Ephemeral sent to client: Csyq2/Schcnk22GwQfMKe4i6qCKIX7kbPAwdV6O0lTs=
2024-03-01T20:07:36.565	Debug	ZkProof client session hash received: pMADV8W2+UK5neRD+V9jszeb/5yoIxj4dM3N1UyA7Mw=
2024-03-01T20:07:36.565	Debug	ZkProof session hash sent to client: wHGs9h88Yg4JWO0s6g8IqLoHe7d5Zq4ncPe+Yi0qxlE=
2024-03-01T20:07:36.735	Debug	Encrypted data received from server:  [here comes a lot of data ...]
2024-03-01T20:07:36.758	Debug	Decrypted data received from server:  [... and presumably the data in decrypted form] 
2024-03-01T20:07:36.767	Error	Error in ProcessRequest: System.NotSupportedException: Serialization and deserialization of 'System.Reflection.MethodBase' instances is not supported. Path: $.TargetSite.:: ---> System.NotSupportedException: Serialization and deserialization of 'System.Reflection.MethodBase' instances is not supported.::   at System.Text.Json.Serialization.Converters.UnsupportedTypeConverter`1.Write(Utf8JsonWriter writer, T value, JsonSerializerOptions options)::   at System.Text.Json.Serialization.JsonConverter`1.TryWrite(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)::   at System.Text.Json.Serialization.Metadata.JsonPropertyInfo`1.GetMemberAndWriteJson(Object obj, WriteStack& state, Utf8JsonWriter writer)::   at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1.OnTryWrite(Utf8JsonWriter writer, T value, JsonSerializerOptions options, WriteStack& state)::   at System.Text.Json.Serialization.JsonConverter`1.TryWrite(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)::   at System.Text.Json.Serialization.JsonConverter`1.WriteCore(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)::   --- End of inner exception stack trace ---::   at System.Text.Json.ThrowHelper.ThrowNotSupportedException(WriteStack& state, NotSupportedException ex)::   at System.Text.Json.Serialization.JsonConverter`1.WriteCore(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)::   at System.Text.Json.Serialization.Metadata.JsonTypeInfo`1.Serialize(Utf8JsonWriter writer, T& rootValue, Object rootValueBoxed)::   at System.Text.Json.Serialization.Metadata.JsonTypeInfo`1.SerializeAsObject(Utf8JsonWriter writer, Object rootValue)::   at System.Text.Json.JsonSerializer.WriteBytesAsObject(Object value, JsonTypeInfo jsonTypeInfo)::   at ServiceWire.DefaultSerializer.Serialize(Object obj, String typeConfigName)::   at ServiceWire.ParameterTransferHelper.SendParameters(Boolean useCompression, Int32 compressionThreshold, BinaryWriter writer, Object[] parameters)::   at ServiceWire.Host.ProcessInvocation(ZkSession session, BinaryReader binReader, BinaryWriter binWriter, Stopwatch sw)::   at ServiceWire.Host.ProcessRequest(Stream readStream, Stream writeStream)

The stats file looks like this:

<entry fr="2024-03-01T19:59:10.433" to="2024-03-01T20:22:22.147" cnt="61" secs="1391,714101">
  <cat nm="ICongressUI" cnt="61">
    <stat nm="SetCapturedLastSlideBitmap" cnt="30" tot="10635" avg="354,5" />
    <stat nm="SetCapturedFirstSlideBitmap" cnt="31" tot="15326" avg="494,3871" />
  </cat>
</entry>

I ran my test several times and stats always shows cnt="30" and cnt="31".

And again, an alternative .NET 7 version using Tcp ran without issues.

**Comment by tylerje** [comment]
That's very strange. What happens if you don't use Zk security? No username and password required.

And is the client a separate process for each call or every two calls? And are you capturing the count on host side or client side?

**Comment by sbuethe** [comment]
The client is a PowerPoint add-in which runs the RPC call on the beginning of a slideshow and at the end - right before PowerPoint closes and the process is destroyed. So it is two calls for each instance.

I will try without zk.

**Comment by tylerje** [comment]
Ah. That is interesting. So it likely runs in a limited sandbox runtime. Is this PP running on a single machine or multiple machines? I'm assuming single machine since you were using Named Pipes. Have you tested this outside the runtime in PP?

**Comment by sbuethe** [comment]
Not with Tcp, but the Named Pipes test I performed with a .NET 8 client was done with a WPF app without any Sandbox. And at least with NamedPipes, it failed the same way.

**Comment by sbuethe** [comment]
As for capturing the number of successful calls: My host app has a log and the number of successful calls matches with the count in the stats file. It always seems to be call #62 that is failing.

**Comment by tylerje** [comment]
Something on .NET 8 seems to be broken. I'll have to set up some test, but it may be some time before I have availability.

**Comment by sbuethe** [comment]
No problem, I do not have any specific reason to use .NET 8 at the moment. I can use .NET 7 in the meantime.

If I find some time, I will try to reproduce the issue with very basic versions of my host and client apps, which I could then share.

**Comment by sbuethe** [comment]
Fyi, the next test run without zk just failed in the same way...

**Comment by sbuethe** [comment]
[RPC Test WPF App (Client to Host Image Transfer).zip](https://github.com/tylerjensen/ServiceWire/files/14735940/RPC.Test.WPF.App.Client.to.Host.Image.Transfer.zip)

After extensive troubleshooting I can confirm that this was a false alarm - I'm sorry.

After I added the ServiceWire source code to my project, I was able to eventually pin point this to an exception which occured only after invoking my code. 

It took some time to find the cause because (as already mentioned) the problem only occurred in .NET 8 and only when code optimizations were turned on - and even then only after many calls. No exception was thrown in debug mode, nor did it occur when try-catch blocks were placed around the code. Ultimately it was a stupid mistake in my code in retrieving stack frames for diagnostic purposes.

Anyway, I can now confirm that ServiceWire works very well with .NET 8 - at least in my use case.

Thank you again for your support and sorry for the trouble.

P.S.: I am attaching my simple WPF test host and client that I put together to try out my specific use case with new client instances being launched by the host. Maybe it's of use for something.
